horovod/mxnet/mpi_ops.py

```diff
@@ -71,13 +72,19 @@ def allreduce(tensor, average=True, name=None):
                          dtype=tensor.dtype)
     c_in = tensor.handle
     c_out = output.handle
+    mx_handle = ctypes.c_int()
+
     if isinstance(name, string_types):
-        check_call(MPI_MXNET_LIB_CTYPES.horovod_mxnet_allreduce_async(c_in,
-                   c_out, c_str(name), ctypes.c_bool(average)))
+        check_call(MPI_MXNET_LIB_CTYPES.horovod_mxnet_allreduce_async(
+            c_in, c_out, c_str(name), ctypes.c_bool(average),
+            ctypes.byref(mx_handle)))
     else:
-        check_call(MPI_MXNET_LIB_CTYPES.horovod_mxnet_allreduce_async(c_in,
-                   c_out, name, ctypes.c_bool(average)))
-    return output
+        check_call(MPI_MXNET_LIB_CTYPES.horovod_mxnet_allreduce_async(
+            c_in, c_out, name, ctypes.c_bool(average),
+            ctypes.byref(mx_handle)))
+
+    _handle_map[mx_handle.value] = output
+    return synchronize(mx_handle.value)
```

Comment:
Referring to these numbers:
```
# gpus | Without HA |   With HA
---------------------------------
   8   |  3072  (NA)|  3078  (NA)
  16   |  6027 (98%)|  5859 (95%)
  32   | 12030 (98%)| 11675 (95%)
  64   | 22346 (83%)| 23166 (94%)
 128   | 40938 (84%)| 45972 (93%)
 256   | 64998 (66%)| 89858 (91%)
```

Reply:
Yes, it is possible to remove the `synchronize()` and reply on MXNet engine to handle the dependencies between tasks.  We plan to further improve the performance after we merge the current stable PR into Horovod. Does that sound good to you?

Reply:
Yep, sounds good.  Do you have a sense of scaling efficiency we'll see with the current version?  I hope to include MXNet support in upcoming 0.16.0 release next week, and I wanted to see if we can publish good scaling #s with it.

Reply:
@alsrgv We found the throughput is affected when we use `synchronize()`. I am reverting to the original implementation now. However, we need a better mechanism to catch the error status returned by MPI. I am currently trying to introduce a `context` variable to store the `Status` inside the callback just like Tensorflow. If you have other better suggestion, it will be greatly appreciated.

Reply:
I think if MXNet has a mechanism to notify framework about op failure, which it will propagate to the user, it would be the best option to use.